### PR TITLE
feat: sidebar UX redesign — attention engine + visual indicators

### PR DESCRIPTION
**Pull request by donovan-yohan**
## Summary

- **Display state machine**: Extended with `needs-answer` (◇) and `error` (■) states, wired `permissionType` discriminator through server → WebSocket → frontend
- **SessionIndicator**: Text-character shape language (●▶◆◇■─) with color-alpha pulse animations replacing CSS dot indicators
- **PR glyphs**: Icon-only PR status characters (◌○✕✓●⊘) replacing StatusDot SVGs
- **Read/unread tracking**: Pure `shouldMarkUnread` logic + reactive localStorage-persisted unread state with bold name + accent border
- **Attention scoring**: `computeAttentionScore` sorts sessions by state priority, PR urgency, recency, and unread status; workspaces sort by highest-scoring session
- **Summary pips**: Collapsed workspace headers show state-summary pips (e.g. ◆1 ●2 ─1) ordered by urgency
- **Branch WebSocket events**: `session-branch-changed` broadcast with CipherText decode animation
- **Stale repo names**: Workspace names derived from git remote instead of directory basename
- **Double-click**: Workspace header double-click to expand/collapse

## Test Plan
- [ ] Verify `npm test` passes (590/591, 1 pre-existing flaky GitWatcher)
- [ ] Verify SessionIndicator shapes render for each display state
- [ ] Verify collapsed workspace headers show summary pips
- [ ] Verify attention sorting reorders sessions by urgency
- [ ] Verify unread bold + accent border appears on state transitions
- [ ] Verify PR glyphs show correct characters and colors
- [ ] Verify double-click workspace header toggles collapse

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: frontend/src/lib/state/sidebar-items.ts [hunk 1]
```diff
@@ -1,6 +1,7 @@
 import type { SessionSummary, WorktreeInfo, Repo, SidebarItem } from '../types.js';
 import type { BackendDisplayState, DisplayState } from './display-state.js';
 import { transitionDisplayState } from './display-state.js';
+import { sortByAttention } from './attention.js';
```
Comment (donovan-yohan) [inline frontend/src/lib/state/sidebar-items.ts:4]: **[code-reviewer]** 🔴 **Critical: `isUnread` state lost on sidebar rebuild**

`buildSidebarItems()` creates fresh `SidebarItem` objects (at lines 130-141, 150-161, 168-179, 191-202) without restoring `isUnread` from existing items or from the `isUnread()` function in `unread.svelte.ts`. When `refreshAll()` runs, all `item.isUnread` flags reset to `undefined`.

Impact:
- Unread indicators disappear after any full refresh
- Attention scores drop by 300 points, changing sort order
- localStorage persistence gives a false sense of durability

**Fix:** Add an import here and restore `isUnread` in each `result.push`:
```typescript
import { isUnread as checkUnread } from './unread.svelte.js';
// ... in each result.push:
isUnread: checkUnread(groupPath),
```

File: frontend/src/lib/state/display-state.ts [hunk 2]
```diff
@@ -16,11 +17,15 @@ export function transitionDisplayState(current: DisplayState, event: DisplayEven
         case 'running':
           return 'running';
         case 'permission':
+          if (event.permissionType === 'question') return 'needs-answer';
           return 'permission';
+        case 'error':
+          return 'error';
         case 'initializing':
           return 'initializing';
       }
```
Comment (donovan-yohan) [inline frontend/src/lib/state/display-state.ts:28]: **[silent-failure-hunter, code-reviewer, code-simplifier, type-design-analyzer, comment-analyzer]** 🟠 **High: Switch fallthrough — `backend-state-changed` falls into `user-viewed`**

Flagged by **5 of 6** review agents. The inner switch covers all 5 `BackendDisplayState` variants and each returns, so fallthrough never occurs *today*. But if a new variant is added without updating the inner switch, execution silently falls through to `user-viewed`, clearing `unseen-idle` or `permission` state.

The `eslint-disable-next-line no-fallthrough` on line 28 suppresses the warning without fixing the structure.

**Fix:** Add a defensive return after the inner switch:
```typescript
  }
  return current; // exhaustive today, safe if BackendDisplayState grows
}
```
Remove the eslint-disable comment.

File: frontend/src/lib/state/unread.svelte.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
+const UNREAD_STORAGE_KEY = 'claude-remote-unread';
+
+let unreadItems = $state<Set<string>>(loadUnread());
+
+function loadUnread(): Set<string> {
+  try {
+    const stored = localStorage.getItem(UNREAD_STORAGE_KEY);
+    return stored ? new Set(JSON.parse(stored)) : new Set();
+  } catch { return new Set(); }
```
Comment (donovan-yohan) [inline frontend/src/lib/state/unread.svelte.ts:12]: **[silent-failure-hunter]** 🟠 **High: Empty catch swallows JSON parse errors, silently loses unread state**

This catches both "localStorage unavailable" (fine) and "corrupted JSON" (data loss). Also, `JSON.parse` returning a non-array (e.g., `"hello"`) creates a set of characters, not IDs.

**Fix:** Validate the parsed structure:
```typescript
function loadUnread(): Set<string> {
  try {
    const stored = localStorage.getItem(UNREAD_STORAGE_KEY);
    if (!stored) return new Set();
    const parsed = JSON.parse(stored);
    if (!Array.isArray(parsed)) return new Set();
    return new Set(parsed.filter(id => typeof id === 'string'));
  } catch {
    return new Set();
  }
}
```

File: frontend/src/lib/state/unread.svelte.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+export function markRead(itemId: string): void {
+  if (unreadItems.has(itemId)) {
+    unreadItems.delete(itemId);
+    saveUnread();
+  }
+}
+
+export function pruneUnread(validIds: Set<string>): void {
```
Comment (donovan-yohan) [inline frontend/src/lib/state/unread.svelte.ts:39]: **[code-simplifier, type-design-analyzer]** 🟡 **Medium: `pruneUnread` exported but never called — localStorage leak**

Stale IDs accumulate in localStorage indefinitely as sessions are created and destroyed.

**Fix:** Call `pruneUnread` in `refreshAll()` after rebuilding sidebar items, passing the set of current item IDs.

File: frontend/src/lib/state/attention.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+  'unseen-idle':   500,
+  running:         100,
+  initializing:     50,
+  'seen-idle':      10,
+  inactive:          1,
+};
+
+function minutesSinceLastActivity(item: SidebarItem): number {
+  if (!item.lastActivity) return Infinity;
+  return (Date.now() - new Date(item.lastActivity).getTime()) / 60_000;
```
Comment (donovan-yohan) [inline frontend/src/lib/state/attention.ts]: **[silent-failure-hunter, pr-test-analyzer]** 🟡 **Medium: Invalid date string produces `NaN`, corrupts entire sort order**

If `item.lastActivity` is a non-empty but malformed string, `new Date("invalid").getTime()` returns `NaN`. Then `score += NaN` makes the entire attention score `NaN`, and `.sort()` produces undefined order.

**Fix:**
```typescript
function minutesSinceLastActivity(item: SidebarItem): number {
  if (!item.lastActivity) return Infinity;
  const ms = new Date(item.lastActivity).getTime();
  if (Number.isNaN(ms)) return Infinity;
  return (Date.now() - ms) / 60_000;
}
```

File: frontend/src/lib/state/sidebar-items.ts [hunk 6]
```diff
@@ -12,7 +13,8 @@ import { transitionDisplayState } from './display-state.js';
 function sessionToBackendState(session: SessionSummary): BackendDisplayState {
   const { agentState, idle } = session;
   if (agentState === 'permission-prompt') return 'permission';
-  if (agentState === 'processing' || agentState === 'error') return 'running';
+  if (agentState === 'error') return 'error';
```
Comment (donovan-yohan) [inline frontend/src/lib/state/sidebar-items.ts:16]: **[comment-analyzer]** 🟡 **Medium: JSDoc priority order missing `error`**

The JSDoc at line 8 says "Priority order: permission > running > initializing > idle" but the code now includes `error` at priority 3. Same issue at line 28.

**Fix:** Update both JSDoc comments to: `permission > error > running > initializing > idle`

File: frontend/src/components/WorkspaceItem.svelte [hunk 7]
```diff
@@ -52,10 +54,46 @@
...
+    'unseen-idle':  { char: '▶', colorClass: 'pip-yellow' },
+    running:        { char: '●', colorClass: 'pip-green' },
+    initializing:   { char: '●', colorClass: 'pip-green' },
+    'seen-idle':    { char: '▶', colorClass: 'pip-yellow-muted' },
+    inactive:       { char: '─', colorClass: 'pip-gray' },
+  };
+
+  const urgencyOrder = ['permission', 'needs-answer', 'error', 'unseen-idle', 'running', 'initializing', 'seen-idle', 'inactive'];
+
+  let summaryPips = $derived((): StatePip[] => {
```
Comment (donovan-yohan) [inline frontend/src/components/WorkspaceItem.svelte]: **[code-reviewer, code-simplifier]** 🟡 **Medium: `$derived` wrapping a function instead of `$derived.by`**

```typescript
let summaryPips = $derived((): StatePip[] => { ... });
```

This creates a reactive *function*, requiring `summaryPips()`. The idiomatic Svelte 5 pattern for multi-line computed values is `$derived.by`:

```typescript
let summaryPips = $derived.by((): StatePip[] => { ... });
```

Then use `summaryPips` directly instead of `summaryPips()`.

File: frontend/src/components/WorkspaceItem.svelte [hunk 8]
```diff
@@ -52,10 +54,46 @@
...
-    const state = sidebarItemById.get(groupPath)?.displayState ?? 'inactive';
-    return 'status-dot status-dot--' + state;
-  }
+  interface StatePip {
+    char: string;
+    colorClass: string;
+    count: number;
+  }
+
+  const pipConfig: Record<string, { char: string; colorClass: string }> = {
```
Comment (donovan-yohan) [inline frontend/src/components/WorkspaceItem.svelte]: **[silent-failure-hunter, type-design-analyzer]** 🟡 **Medium: `pipConfig` typed as `Record<string, ...>` instead of `Record<DisplayState, ...>`**

TypeScript won't catch missing keys if a new `DisplayState` variant is added. Compare with `STATE_SCORES` in `attention.ts` which correctly uses `Record<DisplayState, number>`.

**Fix:** `const pipConfig: Record<DisplayState, { char: string; colorClass: string }> = { ... }`

File: frontend/src/components/Sidebar.svelte [hunk 9]
```diff
@@ -87,9 +88,22 @@
   let mobileDragEnabled = $state(false);
   let dragDisabled = $derived(isTouchDevice && !mobileDragEnabled);
 
+  // Track whether user has manually reordered in this session
+  let userHasDragged = $state(false);
+
+  // Attention-sorted workspace list
+  let attentionSortedRepos = $derived(
+    [...sessionState.repos].sort((a, b) => {
+      const aItems = sessionState.sidebarItems.filter(i => i.repoPath === a.path);
```
Comment (donovan-yohan) [inline frontend/src/components/Sidebar.svelte]: **[code-simplifier]** 🟡 **Medium: O(n·m·log m) filter inside sort comparator**

The sort comparator calls `.filter()` on `sidebarItems` for every comparison. Pre-compute a `Map<repoPath, SidebarItem[]>` once:

```typescript
let itemsByRepo = $derived(new Map(
  sessionState.repos.map(r => [
    r.path,
    sessionState.sidebarItems.filter(i => i.repoPath === r.path)
  ])
));
```

File: frontend/src/lib/ws.ts [hunk 10]
```diff
@@ -1,12 +1,13 @@
 import type { Terminal } from '@xterm/xterm';
-
+import type { BackendDisplayState } from './state/display-state.js';
 const wsProtocol = location.protocol === 'https:' ? 'wss:' : 'ws:';
 
 interface EventMessage {
```
Comment (donovan-yohan) [inline frontend/src/lib/ws.ts:5]: **[type-design-analyzer]** 🟡 **Medium: `EventMessage` is a bag-of-optionals — highest-leverage type improvement**

All fields except `type` are optional, and `type` is an untyped `string`. `branch` and `branchName` are both present, used by different events — confusing.

**Recommended:** Refactor into a discriminated union for compile-time safety.

File: frontend/src/lib/state/sessions.svelte.ts [hunk 11]
```diff
@@ -105,7 +107,19 @@ export function renameSession(sessionId: string, branchName: string, displayName
   }
 }
 
-export function handleBackendStateChanged(sessionId: string, backendState: BackendDisplayState): void {
+export function handleBranchChanged(sessionId: string, branch: string): void {
```
Comment (donovan-yohan) [inline frontend/src/lib/state/sessions.svelte.ts:157]: **[silent-failure-hunter]** 🟡 **Medium: `handleBranchChanged` silently drops updates for unknown sessions**

Both lookups silently no-op on miss. If a WebSocket event arrives before `refreshAll` populates the session list (timing race), the branch update is dropped and the CipherText animation never fires.

Consider logging at debug level when session is not found.

File: frontend/src/lib/state/attention.ts [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+export function computeAttentionScore(item: SidebarItem): number {
+  let score = STATE_SCORES[item.displayState] ?? 0;
```
Comment (donovan-yohan) [inline frontend/src/lib/state/attention.ts]: **[silent-failure-hunter]** ℹ️ **Low: `?? 0` fallback masks unknown states**

`STATE_SCORES` is `Record<DisplayState, number>` so every key exists. The `?? 0` will never execute. If `DisplayState` is extended without updating `STATE_SCORES`, the fallback silently assigns score 0 (lower than `inactive` at 1). Consider removing `?? 0` since the Record type already guarantees completeness.

File: frontend/src/lib/state/unread.svelte.ts [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+  } catch { return new Set(); }
+}
+
+function saveUnread(): void {
+  try {
+    localStorage.setItem(UNREAD_STORAGE_KEY, JSON.stringify([...unreadItems]));
+  } catch { /* localStorage unavailable */ }
+}
+
+export function isUnread(itemId: string): boolean {
```
Comment (donovan-yohan) [inline frontend/src/lib/state/unread.svelte.ts:21]: **[code-simplifier]** ℹ️ **Low: `isUnread()` exported but never called**

The read path uses `item.isUnread` on `SidebarItem`. This is dead code unless wired into `buildSidebarItems` (where importing this function is the recommended fix for the critical isUnread state loss bug).

File: frontend/src/lib/state/sessions.svelte.ts [hunk 14]
```diff
@@ -128,10 +143,22 @@ export function handleBackendStateChanged(sessionId: string, backendState: Backe
   item.lastKnownBackendState = backendState;
 
   // Apply transition
-  const newDisplayState = transitionDisplayState(item.displayState, { type: 'backend-state-changed', state: backendState });
+  const newDisplayState = transitionDisplayState(
+    item.displayState,
+    permissionType
```
Comment (donovan-yohan) [inline frontend/src/lib/state/sessions.svelte.ts:199]: **[code-simplifier]** ℹ️ **Low: Unnecessary ternary for optional `permissionType`**

`DisplayEvent` marks `permissionType` as optional. Passing `undefined` is equivalent to omitting it:
```typescript
const newDisplayState = transitionDisplayState(
  item.displayState,
  { type: 'backend-state-changed', state: backendState, permissionType },
);
```

**Review by donovan-yohan — COMMENTED** [review]
## PR Review Summary

**Agents run:** code-reviewer, silent-failure-hunter, pr-test-analyzer, type-design-analyzer, comment-analyzer, code-simplifier

### Key Findings

**2 Critical bugs** (state loss on rebuild), **1 High-risk latent defect** (switch fallthrough), plus improvements across error handling, type safety, test coverage, and code style.

| Severity | Count |
|----------|-------|
| Critical | 2 |
| High | 5 |
| Medium | 8 |
| Low | 3 |

### Critical Finding Not Mapped to Diff Line

**`sidebar-items.ts:231` — `needs-answer` state lost on reconciliation** (code-reviewer)

`reconcileDisplayState` calls `transitionDisplayState` without `permissionType`. After `refreshAll()` or page reload, a `needs-answer` session (hollow diamond ◇) incorrectly shows as `permission` (filled diamond ◆). Fix: preserve `needs-answer` when backend still says `permission`:
```typescript
if (newBackendState === 'permission' && existing.displayState === 'needs-answer') {
  return 'needs-answer';
}
```

### ADR Compliance

0 architecture rule violations found. All invariants maintained.

### Positive Highlights

- Excellent pure logic / reactive wrapper separation — trivially testable with node:test
- `Record<DisplayState, number>` pattern in STATE_SCORES is the gold standard for type-safe lookup tables
- Display state transition table test is thorough with 20 explicit cases
- PR status priority tests cover critical edge cases (merged > all, draft > changes-requested)
- `shouldMarkUnread` / `shouldNotify` predicates avoid duplicating attention-state lists

🤖 Generated with [Claude Code](https://claude.com/claude-code)

File: test/pr-status.test.ts [hunk 15]
```diff
@@ -61,3 +62,20 @@ describe('derivePrDotStatus', () => {
...
   });
 });
+
+describe('prGlyph', () => {
+  it('maps each status to a unique character', () => {
+    const statuses: PrDotStatus[] = ['draft', 'open', 'review-requested', 'changes-requested', 'approved', 'merged', 'closed', 'unknown'];
+    const chars = statuses.map(s => prGlyph(s).char);
+    for (const c of chars) {
+      assert.ok(c.length > 0);
+    }
```
[the remark this answers is not in the document]
Comment (donovan-yohan) [reply test/pr-status.test.ts:72]: Fixed — tightened the assertion to enforce full `Set` uniqueness across all statuses. Also changed `review-requested` glyph from `○` to `◎` so all 8 glyphs are distinct.

File: frontend/src/lib/state/display-state.ts [hunk 16]
```diff
@@ -16,11 +17,15 @@ export function transitionDisplayState(current: DisplayState, event: DisplayEven
...
     }
+    // eslint-disable-next-line no-fallthrough
```
[the remark this answers is not in the document]
Comment (donovan-yohan) [reply frontend/src/lib/state/display-state.ts]: Fixed — added a `default: return current` case inside the inner switch and removed the eslint-disable comment.

File: frontend/src/lib/state/sessions.svelte.ts [hunk 17]
```diff
@@ -128,10 +143,22 @@ export function handleBackendStateChanged(sessionId: string, backendState: Backe
...
+  );
   if (newDisplayState !== oldDisplayState) {
     item.displayState = newDisplayState;
 
+    // Track unread
+    const isViewing = item.sessions.some(s => s.id === sessionId && s.id === activeSessionId);
+    if (shouldMarkUnread(oldDisplayState, newDisplayState, isViewing)) {
+      markUnread(item.id);
+      item.isUnread = true;
+    }
```
[the remark this answers is not in the document]
Comment (donovan-yohan) [reply frontend/src/lib/state/sessions.svelte.ts]: Fixed — simplified the check to `item.sessions.some(s => s.id === activeSessionId)` which correctly checks if the user is viewing *any* session in the group.

File: frontend/src/lib/state/sidebar-items.ts [hunk 18]
```diff
@@ -198,7 +202,7 @@ export function buildSidebarItems(
     });
   }
```
[the remark this answers is not in the document]
Comment (donovan-yohan) [reply frontend/src/lib/state/sidebar-items.ts:209]: Fixed — `buildSidebarItems` now accepts an optional `checkUnread` callback (injected by the Svelte caller) and restores `isUnread` from existing items or localStorage on every rebuild. prStatus enrichment is deferred to Phase 2 as planned.

File: frontend/src/components/Sidebar.svelte [hunk 19]
```diff
@@ -87,9 +88,22 @@
...
+      const bItems = sessionState.sidebarItems.filter(i => i.repoPath === b.path);
```
[the remark this answers is not in the document]
Comment (donovan-yohan) [reply frontend/src/components/Sidebar.svelte]: prStatus enrichment from `orgPrs` is a Phase 2 item — the attention scorer handles `undefined` prStatus gracefully (no PR score added). Pre-computed `itemsByRepo` map has been added to fix the O(n·m·log m) sort performance issue flagged separately.

File: server/index.ts [hunk 20]
```diff
@@ -313,6 +312,7 @@ async function main(): Promise<void> {
         }
       }
     }
+    broadcastBranchChanged(cwdPath, newBranch);
     // Rebuild ref watchers when branches change (new upstream to watch)
```
[the remark this answers is not in the document]
Comment (donovan-yohan) [reply server/index.ts:318]: Fixed — now matches sessions by `worktreePath ?? repoPath` instead of `session.cwd`, which correctly handles sessions that have changed directory during their lifetime.

File: test/worktrees.test.ts [hunk 21]
```diff
@@ -454,6 +454,20 @@ describe('mountain name collision retry', () => {
...
+  it('derives repo name from git remote origin URL', async () => {
+    const { execFile } = await import('node:child_process');
+    const { promisify } = await import('node:util');
+    const exec = promisify(execFile);
+
+    const result = await exec('git', ['remote', 'get-url', 'origin'], { cwd: process.cwd() });
+    const url = result.stdout.trim();
+    const name = url.split('/').pop()?.replace(/\.git$/, '') ?? '';
+    assert.ok(name.length > 0, 'should extract a non-empty name from git remote');
+    assert.ok(!name.includes('/'), 'name should not contain slashes');
```
[the remark this answers is not in the document]
Comment (donovan-yohan) [reply test/worktrees.test.ts]: Fixed — extracted a pure `repoNameFromRemoteUrl(url)` helper in `server/workspaces.ts` and replaced the environment-dependent test with hermetic URL fixture tests.

File: frontend/src/lib/state/sidebar-items.ts (hunk 1 above)
Comment (donovan-yohan) [reply frontend/src/lib/state/sidebar-items.ts:4]: Fixed — `buildSidebarItems` now accepts an optional `checkUnread` callback from the Svelte caller and restores `isUnread` from existing items or localStorage on every rebuild.

File: frontend/src/lib/state/display-state.ts (hunk 2 above)
Comment (donovan-yohan) [reply frontend/src/lib/state/display-state.ts:28]: Fixed — replaced eslint-disable with a `default: return current` case in the inner switch.

File: frontend/src/lib/state/unread.svelte.ts (hunk 3 above)
Comment (donovan-yohan) [reply frontend/src/lib/state/unread.svelte.ts:12]: Fixed — `loadUnread` now validates the parsed value is an array and filters to string-only entries.

File: frontend/src/lib/state/unread.svelte.ts (hunk 4 above)
Comment (donovan-yohan) [reply frontend/src/lib/state/unread.svelte.ts:39]: Fixed — `pruneUnread` is now called in `refreshAll()` after rebuilding sidebar items, passing the set of current item IDs.

File: frontend/src/lib/state/attention.ts (hunk 5 above)
Comment (donovan-yohan) [reply frontend/src/lib/state/attention.ts]: Fixed — `minutesSinceLastActivity` now checks for `NaN` and returns `Infinity` for malformed dates.

File: frontend/src/lib/state/sidebar-items.ts (hunk 6 above)
Comment (donovan-yohan) [reply frontend/src/lib/state/sidebar-items.ts:16]: Fixed — both JSDoc comments now read `permission > error > running > initializing > idle`.

File: frontend/src/components/WorkspaceItem.svelte (hunk 7 above)
Comment (donovan-yohan) [reply frontend/src/components/WorkspaceItem.svelte]: Fixed — changed to `$derived.by` and updated the usage site to access `summaryPips` directly instead of calling `summaryPips()`.

File: frontend/src/components/WorkspaceItem.svelte (hunk 8 above)
Comment (donovan-yohan) [reply frontend/src/components/WorkspaceItem.svelte]: Fixed — typed as `Record<DisplayState, { char: string; colorClass: string }>`.

File: frontend/src/components/Sidebar.svelte (hunk 9 above)
Comment (donovan-yohan) [reply frontend/src/components/Sidebar.svelte]: Fixed — pre-computed `itemsByRepo` map via `$derived` so the sort comparator does O(1) lookups.

File: frontend/src/lib/state/sessions.svelte.ts (hunk 11 above)
Comment (donovan-yohan) [reply frontend/src/lib/state/sessions.svelte.ts:157]: Fixed — added `console.debug` logging when session or sidebar item is not found for a branch change event.

File: frontend/src/lib/state/attention.ts (hunk 12 above)
Comment (donovan-yohan) [reply frontend/src/lib/state/attention.ts]: Fixed — removed the `?? 0` fallback since `Record<DisplayState, number>` guarantees completeness.

File: frontend/src/lib/state/unread.svelte.ts (hunk 13 above)
Comment (donovan-yohan) [reply frontend/src/lib/state/unread.svelte.ts:21]: No longer dead code — `isUnread` is now imported by `sessions.svelte.ts` and passed to `buildSidebarItems` as the `checkUnread` callback.

File: frontend/src/lib/state/sessions.svelte.ts (hunk 14 above)
Comment (donovan-yohan) [reply frontend/src/lib/state/sessions.svelte.ts:199]: Kept the ternary form due to `exactOptionalPropertyTypes` — passing `undefined` for an optional property is a type error. The ternary conditionally includes `permissionType` only when defined.

File: frontend/src/lib/ws.ts (hunk 10 above)
Comment (donovan-yohan) [reply frontend/src/lib/ws.ts:5]: Acknowledged — deferring the `EventMessage` discriminated union refactor to a follow-up since it touches the entire WS message handling chain across both frontend and server. The current bag-of-optionals works correctly for all current event types.
